SampleShared/SharedAssets/Scripts/PrefabMonoBehaviour.cs

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.OpenXR.Sample
+{
+    public abstract class PrefabMonoBehaviour : MonoBehaviour
+    {
+        private bool m_childrenDirty = false;
+
+        protected void SetPropertyValue<T>(ref T member, T value) where T : IEquatable<T>
+        {
+            if (member.Equals(value))
+            {
+                member = value;
+                m_childrenDirty = true;
+            }
+        }
+
+        protected void SetPropertyObject<T>(ref T member, T value) where T : class
+        {
+            if (member == value)
+            {
+                member = value;
+                m_childrenDirty = true;
+            }
+        }
+
+        protected virtual void InitializeComponents() { }
+        protected virtual void UpdateChildren() { }
+
+        protected void UpdateChidrenWhenDirty()
+        {
+            if (m_childrenDirty)
+            {
+                m_childrenDirty = false;
+                UpdateChildren();
+            }
+        }
+
+        void Start()
```

Comment:
Since this is an abstract base class, we should probably make these `protected` (and maybe `protected virtual`). That way, any derived classes will properly flag if these are being redefined and hidden (intentionally or not). I've hit some bugs in the past where derived MonoBehaviour-based classes will accidentally hide the base implementation (and it never gets called) because `Start`/`Update` etc were marked `private`.

https://answers.unity.com/questions/147276/why-is-start-from-my-base-class-ignored-in-my-sub.html

Reply:
Good to learn this.  Will look into making them virtual protected then.  If so I probably can factor other functions a bit differently as well.